Introduce data schema to store raw tensors

lucylq opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #6540

Differential Revision: [D65156641](https://our.internmc.facebook.com/intern/diff/D65156641)

lucylq committed:
Introduce data schema to store raw tensors

[ghstack-poisoned]

lucylq commented:
See: https://github.com/pytorch/executorch/issues/6492

lucylq committed:
Update on "Introduce data schema to store raw tensors"




[ghstack-poisoned]

lucylq commented on schema/data.fbs:
```diff
@@ -0,0 +1,83 @@
+  // Offset in scalar_type elements (e.g., multiples of 4 bytes for an int
+  // scalar type) from the beginning of the tensor buffer to the beginning of
+  // the actual data. Currently, the runtime only supports a value of zero.
+  storage_offset:int;
```
`storage_offset` and `layout` taken from `program.fbs`, don't think there is an immediate use case for them.

lucylq commented on schema/data.fbs:
```diff
@@ -0,0 +1,83 @@
+  scalar_type:ScalarType;
+
+  // Size of each dimension.
+  dim_sizes:[int];
```
equivalent of 'sizes' in `program.fbs`, using `dim_sizes` to differentiate from `size` of the tensor below.

lucylq commented:
@lucylq has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D65156641).

lucylq committed:
Update on "Introduce data schema to store raw tensors"


Differential Revision: [D65156641](https://our.internmc.facebook.com/intern/diff/D65156641)

[ghstack-poisoned]

lucylq committed:
Update on "Introduce data schema to store raw tensors"


Differential Revision: [D65156641](https://our.internmc.facebook.com/intern/diff/D65156641)

[ghstack-poisoned]

lucylq commented:
@lucylq has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D65156641).

lucylq committed:
Update on "Introduce data schema to store raw tensors"


Differential Revision: [D65156641](https://our.internmc.facebook.com/intern/diff/D65156641)

[ghstack-poisoned]

JacobSzwejbka commented on exir/schema_data.py:
```diff
@@ -0,0 +1,38 @@
+class TensorMetadata:
+    fully_qualified_name: str
+    scalar_type: ScalarType
+    dim_sizes: List[int]
```
just sizes

JacobSzwejbka reviewed (approved):
I was thinking maybe we should try to reuse more fields from the core schema but I guess the goal of this is different enough that we dont have to. Are we 100% sure we even need sizes and dim order. I think the set of fully qualified names provides a good enough compatibility guarantee (seems unlikely 2 models would have the exact same key set but different values)

lucylq committed:
Update on "Introduce data schema to store raw tensors"


Differential Revision: [D65156641](https://our.internmc.facebook.com/intern/diff/D65156641)

[ghstack-poisoned]

lucylq committed:
Update on "Introduce data schema to store raw tensors"


Differential Revision: [D65156641](https://our.internmc.facebook.com/intern/diff/D65156641)

[ghstack-poisoned]

lucylq committed:
Update on "Introduce data schema to store raw tensors"


Differential Revision: [D65156641](https://our.internmc.facebook.com/intern/diff/D65156641)

[ghstack-poisoned]

lucylq committed:
Update on "Introduce data schema to store raw tensors"


Differential Revision: [D65156641](https://our.internmc.facebook.com/intern/diff/D65156641)

[ghstack-poisoned]
